pkg/mimirtool/client/backfill.go

```diff
@@ -0,0 +1,196 @@
+			resp.StatusCode, resp.Status)
+	}
+
+	// Upload each block file
+	for _, tf := range blockMeta.Thanos.Files {
+		if tf.RelPath == block.MetaFilename {
+			// Don't upload meta file in this step
+			continue
+		}
+
+		pth := filepath.Join(blockDir, filepath.FromSlash(tf.RelPath))
+		f, err := os.Open(pth)
+		if err != nil {
+			return errors.Wrapf(err, "failed to open %q", pth)
+		}
+		defer func() {
+			_ = f.Close()
+		}()
+
+		st, err := f.Stat()
+		if err != nil {
+			return errors.Wrapf(err, "failed to get file info for %q", pth)
+		}
+
+		escapedPath := url.QueryEscape(tf.RelPath)
+		level.Info(logger).Log("msg", "uploading block file", "path", pth, "size", st.Size())
+		resp, err := c.doRequest(path.Join(blockPrefix, fmt.Sprintf("files?path=%s", escapedPath)),
+			http.MethodPost, f, st.Size())
+		if err != nil {
+			return errors.Wrapf(err, "request to upload file %q failed", pth)
+		}
+		closeResp(resp)
+		if resp.StatusCode/100 != 2 {
+			return fmt.Errorf("request to upload block file failed, with HTTP status %d %s",
+				resp.StatusCode, resp.Status)
+		}
+	}
+
+	resp, err = c.doRequest(fmt.Sprintf("%s?uploadComplete=true", blockPrefix), http.MethodPost,
+		nil, -1)
+	if err != nil {
+		return errors.Wrap(err, "request to finish block upload failed")
+	}
+	closeResp(resp)
+	if resp.StatusCode/100 != 2 {
+		return fmt.Errorf("request to finish block upload failed, with HTTP status %d %s",
+			resp.StatusCode, resp.Status)
+	}
+
+	level.Info(logger).Log("msg", "block uploaded successfully")
+
+	return nil
+}
+
+func getBlockMeta(blockDir string) (metadata.Meta, error) {
+	var blockMeta metadata.Meta
+
+	metaPath := filepath.Join(blockDir, block.MetaFilename)
+	f, err := os.Open(metaPath)
+	if err != nil {
+		return blockMeta, errors.Wrapf(err, "failed to open %q", metaPath)
+	}
+	defer func() {
+		_ = f.Close()
+	}()
+
+	if err := json.NewDecoder(f).Decode(&blockMeta); err != nil {
+		return blockMeta, errors.Wrapf(err, "failed to decode %q", metaPath)
+	}
+
+	idxPath := filepath.Join(blockDir, block.IndexFilename)
+	idxSt, err := os.Stat(idxPath)
+	if err != nil {
+		return blockMeta, errors.Wrapf(err, "failed to stat %q", idxPath)
+	}
+	blockMeta.Thanos.Files = []metadata.File{
+		{
+			RelPath:   block.IndexFilename,
+			SizeBytes: idxSt.Size(),
+		},
+		{
+			RelPath: block.MetaFilename,
+		},
+	}
+
+	chunksDir := filepath.Join(blockDir, block.ChunksDirname)
+	entries, err := os.ReadDir(chunksDir)
+	if err != nil {
+		return blockMeta, errors.Wrapf(err, "failed to read dir %q", chunksDir)
+	}
+	reChunk := regexp.MustCompile(`^\d{6}$`)
+	for _, e := range entries {
+		if !reChunk.MatchString(e.Name()) {
+			continue
+		}
+
+		pth := filepath.Join(chunksDir, e.Name())
+		st, err := os.Stat(pth)
+		if err != nil {
+			return blockMeta, errors.Wrapf(err, "failed to stat %q", pth)
+		}
+
+		blockMeta.Thanos.Files = append(blockMeta.Thanos.Files, metadata.File{
```

Comment:
if `e.Name()` refers to a directory, we should not include it in meta.json, but report an error.